### ISSUE TITLE
流水并行指定物理硬件问题

### ISSUE THREAD

**Issue by sywang0111** [Oneflow-Inc/oneflow#3996]
machine0&1上各有2个GPU0-1
```
flow.config.gpu_device_num(2)
 nodes = [{"addr": "192.168.1.1"}, {"addr": "192.168.1.2"}]
```
为了实现流水并行将一个全连接层指定到特定的物理硬件。
```
#with flow.scope.placement("gpu", ["1:0"]):
with flow.scope.placement("gpu", ["0:0-1"]):
#with flow.scope.placement("gpu", ["0:0-1","1:0-1"]):
        hidden = flow.layers.dense(...)
```
上面3种方式，指定到machin1的GPU0和所有GPU上均能成功运行，但指定到machine0上的GPU0-1却会报错。
错误信息为：tensor_slice_view.cpp:117] Check failed: axis < num_axes (1 vs. 0)
请问如何解决？

**Comment by jackalcooper** [comment]
> 为了实现流水并行将一个全连接层指定到特定的物理硬件。

可以提供更多一些这个需求的上下文吗

另外能够提供更详细一些的代码，比如work的情况的代码，和不work的情况的代码，最好是能复现问题的

**Comment by sywang0111** [comment]
```
def lenet(data, train=False):
    initializer = flow.truncated_normal(0.1)
    conv1 = flow.layers.conv2d(
        data,
        32,
        5,
        padding="SAME",
        activation=flow.nn.relu,
        kernel_initializer=initializer,
        name="conv1",
        )
    pool1 = flow.nn.max_pool2d(conv1, ksize=2, strides=2, padding="SAME", name="pool1")
    conv2 = flow.layers.conv2d(
        pool1,
        64,
        5,
        padding="SAME",
        activation=flow.nn.relu,
        kernel_initializer=initializer,
        name="conv2",
    )
    pool2 = flow.nn.max_pool2d(conv2, ksize=2, strides=2, padding="SAME", name="pool2")
    reshape = flow.reshape(pool2, [pool2.shape[0], -1])
    #with flow.scope.placement("gpu", ["1:0"]):
    with flow.scope.placement("gpu", ["0:0-1"]):
    #with flow.scope.placement("gpu", ["0:0-1","1:0-1"]):
        hidden = flow.layers.dense(
            reshape,
            512,
            activation=flow.nn.relu,
            kernel_initializer=initializer,
            name="hidden",
            model_distribute=flow.distribute.split(axis=0)###
        )
    if train:
        hidden = flow.nn.dropout(hidden, rate=0.5)

    output = flow.layers.dense(
        hidden, 10, kernel_initializer=initializer, name="outlayer",
        model_distribute=flow.distribute.split(axis=0)
    )
    return output
```
在lenet中将全连接层给指定到物理硬件上，有#指定到machine1上GPU0，和4个GPU上都是可以运行的，中间没有#指定到machine0上GPU0-1的是出错的

**Comment by sywang0111** [comment]
或者说我能在流水并行的情况下，layer能再做数据并行或者模型并行吗

**Comment by jackalcooper** [comment]
> 或者说我能在流水并行的情况下，layer能再做数据并行或者模型并行吗

如果这里的流水并行指的是网络中不同层的op放在不同的机器或者卡上，是可以的。

另外，请发一下完整代码

**Comment by jackalcooper** [comment]
这是我的代码，根据你的代码片段增加的上下文，2机器，每机器2卡。没有复现你的错误
```
# lenet_train.py
import oneflow as flow
import oneflow.typing as tp
import unittest

BATCH_SIZE = 100


def lenet(data, train=False):
    initializer = flow.truncated_normal(0.1)
    conv1 = flow.layers.conv2d(
        data,
        32,
        5,
        padding="SAME",
        activation=flow.nn.relu,
        kernel_initializer=initializer,
        name="conv1",
    )
    pool1 = flow.nn.max_pool2d(conv1, ksize=2, strides=2, padding="SAME", name="pool1")
    conv2 = flow.layers.conv2d(
        pool1,
        64,
        5,
        padding="SAME",
        activation=flow.nn.relu,
        kernel_initializer=initializer,
        name="conv2",
    )
    pool2 = flow.nn.max_pool2d(conv2, ksize=2, strides=2, padding="SAME", name="pool2")
    reshape = flow.reshape(pool2, [pool2.shape[0], -1])
    # with flow.scope.placement("gpu", ["1:0"]):
    with flow.scope.placement("gpu", ["0:0-1"]):
        # with flow.scope.placement("gpu", ["0:0-1","1:0-1"]):
        hidden = flow.layers.dense(
            reshape,
            512,
            activation=flow.nn.relu,
            kernel_initializer=initializer,
            name="hidden",
            model_distribute=flow.distribute.split(axis=0),  ###
        )
    if train:
        hidden = flow.nn.dropout(hidden, rate=0.5)

    output = flow.layers.dense(
        hidden,
        10,
        kernel_initializer=initializer,
        name="outlayer",
        model_distribute=flow.distribute.split(axis=0),
    )
    return output


@flow.unittest.skip_unless_2n2d()
class Test2dGpuVariable(flow.unittest.TestCase):
    def test_2n_lenet(test_case):
        @flow.global_function(type="train")
        def train_job(
            images: tp.Numpy.Placeholder((BATCH_SIZE, 1, 28, 28), dtype=flow.float),
            labels: tp.Numpy.Placeholder((BATCH_SIZE,), dtype=flow.int32),
        ) -> tp.Numpy:
            with flow.scope.placement("gpu", "0:0"):
                logits = lenet(images, train=True)
                loss = flow.nn.sparse_softmax_cross_entropy_with_logits(
                    labels, logits, name="softmax_loss"
                )

            lr_scheduler = flow.optimizer.PiecewiseConstantScheduler([], [0.1])
            flow.optimizer.SGD(lr_scheduler, momentum=0).minimize(loss)
            return loss

        flow.config.gpu_device_num(2)
        check_point = flow.train.CheckPoint()
        check_point.init()

        (train_images, train_labels), (test_images, test_labels) = flow.data.load_mnist(
            BATCH_SIZE, BATCH_SIZE
        )

        for epoch in range(20):
            for i, (images, labels) in enumerate(zip(train_images, train_labels)):
                loss = train_job(images, labels)
                if i % 20 == 0:
                    print(loss.mean())
        check_point.save("./lenet_models_1")  # need remove the existed folder
        print("model saved")


if __name__ == "__main__":
    unittest.main()
```

**Comment by sywang0111** [comment]
...抱歉，应该是环境配置或者其他因素造成的问题，在别的机器上代码就能够运行了。感谢您的热心帮助。
